Added a support for Dataset.rolling.

fujiisoup opened the pull request:
- [x] closes #859
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

There seems to be two approaches to realize Dataset.rolling,
1. Apply rolling in each DataArrays and then combine them.
2. Apply Dataset directoly with some DataArrays that do not depend on `dim` kept aside, then merge them later.

I chose the latter approach to reuse existing `Rolling` object as much as possible,
but it results in some duplicates in `ImplementsRollingDatasetReduce`.
Any feedbacks and comments are very welcome.

MaximilianR commented on xarray/core/dataset.py:
```diff
@@ -307,6 +308,7 @@ class Dataset(Mapping, ImplementsDatasetReduce, BaseDataObject,
     coordinates used for label based indexing.
     """
     groupby_cls = groupby.DatasetGroupBy
+    rolling_cls = rolling.DatasetRolling
```
I think this should be private (I know `groupby_cls` is public, but I think that is a mistake)

MaximilianR commented on xarray/core/rolling.py:
```diff
@@ -190,4 +194,65 @@ class DataArrayRolling(Rolling, ImplementsRollingArrayReduce):
+                                obj.coords)
+        # drop them from obj, then call parent's initializer.
+        obj = obj.drop(self.fixed_ds.data_vars.keys())
+        Rolling.__init__(self, obj, min_periods, center, **windows)
```
Very narrow point, but this should probably be `super`

fujiisoup commented:
@MaximilianR 

Thanks for the review.
I made `rolling_cls` and `groupby_cls` private, and added some docstrings.

Also, I removed `_axis_num` property in Rolling class to make it compatible with Dataset.
It can be replaced by da.get_axis_num(self.dim).

shoyer commented on xarray/core/rolling.py:
```diff
@@ -192,4 +191,76 @@ class DataArrayRolling(Rolling, ImplementsRollingArrayReduce):
+        reduced = super(DatasetRolling, self).reduce(func, **kwargs)
+        # Need to restore dimension order
+        for key, var in reduced.data_vars.items():
+            reduced = reduced.assign(**{key:
```
I think this is unnecessary now, after #1277.

shoyer commented on xarray/core/rolling.py:
```diff
@@ -192,4 +191,76 @@ class DataArrayRolling(Rolling, ImplementsRollingArrayReduce):
+        from .dataset import Dataset
+        dim, window = next(iter(windows.items()))
+        # DataArrays not depending on the rolling-axis are kept aside from obj
+        self.fixed_ds = Dataset({key: da for key, da in obj.data_vars.items()
```
Use an `OrderedDict` here instead of a dict comprehension to preserve the order of `data_vars`.

fujiisoup replied:
I will change the `DatasetRolling` structure to keep `data_var`s order.
My code destroys this order in `reduce` method by calling `ds.merge`.

shoyer commented on xarray/core/common.py:
```diff
@@ -126,6 +169,76 @@ def wrapped_func(self, **kwargs):
+                raise NotImplementedError(
+                    'Rolling window operation does not work with dask arrays')
+
+            result = Dataset({key: (da.dims,
```
Also needs to use `OrderedDict`.

shoyer commented on xarray/core/common.py:
```diff
@@ -70,14 +70,51 @@ def wrapped_func(self, dim=None, keep_attrs=False, **kwargs):
 
 
 class ImplementsRollingArrayReduce(object):
```
I recognize that you didn't write this class, but I wonder if we couldn't simply consolidate it into `DataArrayRolling`? It feels like one more level of complexity than is necessary. The `ImplementsArrayReduce` type classes were implemented because we have multiple array types for which we wanted similar reduce functionality.

Either way, it would be nice to move this logic into the `rolling.py` module, just to keep it in fewer places.

Perhaps @jhamman recalls why we have this separate class...

fujiisoup replied:
Agreed. 
I moved this class into `DataArrayRolling` in the next commit.

shoyer commented on xarray/core/common.py:
```diff
@@ -126,6 +169,76 @@ def wrapped_func(self, **kwargs):
+            else:
+                min_count = self.min_periods
+
+            result = Dataset({key: (da.dims,
```
Needs to use `OrderedDict` here to preserve data variable order.

shoyer commented on doc/whats-new.rst:
```diff
@@ -21,6 +21,8 @@ v0.9.2 (unreleased)
 Enhancements
 ~~~~~~~~~~~~
+- ``rolling`` on Dataset is now supported (:issue:`859`).
+  By `Keisuke Fujii <https://github.com/fujiisoup>`_.
```
Please also add `Dataset.rolling` to `doc/api.rst`.

fujiisoup commented:
@shoyer Thanks for the review.

I didn't notice Dataset keeps the order of `data_vars`.
I think I should change the direction, because the data_var's order is destroyed after the final `merge` call.

Maybe `DatasetRolling` should not be a child of `Rolling`, but an independent class that has an `OrderedDict` of `Rolling` for each `data_var` to preserve the variable order.

fujiisoup commented:
+ Now DatasetRolling.reduce() keeps its original `data_vars` order. (test: )
+ Dataset.rolling is added to api.rst.

I updated `DatasetRolling` so that its `reduce` method relies mostly on `DataArrayRolling.reduce` rather than `window.reduce`.
This change makes not only the reduced-Dataset keep its original order of `data_var`, but also makes the code much cleaner.
I moved many methods from `Rolling` to `DataArrayRolling` because `DatasetRolling` does not use them anymore.

However, I guess `Rolling` class is originally designed to be used also for `Dataset`, as discussed in #859. 
If someone has any requests, I will reconsider the class structure.

fujiisoup commented on xarray/tests/test_dataset.py:
```diff
@@ -3261,3 +3264,132 @@ def test_dir_unicode(data_set):
+    assert_allclose(actual, expected)
+    assert ds.dims == actual.dims
+    # make sure the order of data_var are not changed.
+    assert list(ds.data_vars.keys()) == list(actual.data_vars.keys())
```
Here is the test to make sure that the order of `data_vars` is not changed by `rolling().reduce()`.

shoyer commented:
> I didn't notice Dataset keeps the order of data_vars.

It's not a strict requirement, but we try to do it when possible because it leads to more predictable results.

fujiisoup commented:
>  we try to do it when possible because it leads to more predictable results.

OK. I totally agree with this.
(Today I experienced a bug caused by unexpectedly aligned data_vars in other project. I realized predictable behavior is really important :) )

I think this PR is ready for review. If anyone has comments about the following behavior, I'm happy to update the code.
+ Only data_vars depending on the given rolling-dim are reduced. The others remain unchanged but are still stored in the reduced dataset (not dropped).
+ If rolling-dim is not included in `Dataset.dims`, this class raises `KeyErrror`.
+ The order of data_vars remain unchanged even after `reduce()`.
+ All the attrs are dropped.

jhamman commented:
@fujiisoup - This is looking pretty good. I'll let @MaximilianR give the final review since he's spent more time here than I have. As for your four points above, I think I'm fine with all of those design decisions.

One minor point though (no change needed)

> All the attrs are dropped

I don't think this is true for data_vars that don't have rolling-dim in them.

shoyer commented on xarray/core/rolling.py:
```diff
@@ -236,31 +330,62 @@ def __init__(self, obj, min_periods=None, center=False, **windows):
+        Return a wrapped function for injecting numpy and bottoleneck methods.
+        see ops.inject_datasetrolling_methods
+        """
+        def wrapped_func(self, **kwargs):
```
Can you make a single helper function that takes `func` and `dataset_rolling` as arguments and applies this logic? e.g.,
```
def _reduce_rollings(rolling, func):
    reduced = OrderedDict()
    for key, da in self.obj.data_vars.items():
        if self.dim in da.dims:
            reduced[key] = func(self.rollings[key])
        else:
            reduced[key] = self.obj[key]
    return Dataset(reduced, coords=self.obj.coords)
```
Then both this method and `reduce` can reuse the same logic, just with a different callable (e.g., `lambda x: x.reduce(func, **kwargs)` for reduce).

fujiisoup replied:
@shoyer 
Thank you for the review.
Could you give me more detail about the helper function?
I am wondering how I could distinguish the two types of reducing, 
`rolling.reduce(np.mean)` and `rolling.mean()`, the latter uses bottleneck if it is available.

shoyer replied:
Honestly, maybe this isn't worth the trouble here -- the redundant lines are right next to each other already.

shoyer commented on xarray/core/ops.py:
```diff
@@ -560,3 +560,14 @@ def inject_bottleneck_rolling_methods(cls):
+    for name, f in methods:
+        func = cls._reduce_method(f)
+        func.__name__ = name
+        func.__doc__ =\
```
prefer line breaking after parentheses, if possible, like in `inject_bottleneck_rolling_methods` above.

MaximilianR commented:
> I'll let @MaximilianR give the final review 

Apologies for missing this. I actually think @jhamman has done much more than I to contribute to this area!

But this change looks good, thanks @fujiisoup !

fujiisoup commented:
@shoyer 
Thank you for the reply.
I did not yet think of smarter solution.
I hope the current implementation is acceptable. 
If I come up with cleaner script, I will send another PR.

shoyer commented on xarray/tests/test_dataarray.py:
```diff
@@ -1352,20 +1352,6 @@ def test_reduce_keep_attrs(self):
         self.assertEqual(len(vm.attrs), len(self.attrs))
         self.assertEqual(vm.attrs, self.attrs)
 
-    def test_assign_attrs(self):
```
Did you intend to delete this test?

fujiisoup replied:
It seems my mistake...
Thank you for finding it.

I'll send another PR.

shoyer commented on xarray/tests/test_dataarray.py:
```diff
@@ -2544,4 +2530,3 @@ def test_rolling_reduce(da, center, min_periods, window, name):
     actual = rolling_obj.reduce(getattr(np, 'nan%s' % name))
     expected = getattr(rolling_obj, name)()
     assert_allclose(actual, expected)
-    assert da.dims == expected.dims
```
I don't think there's any reason to delete this line?

fujiisoup replied:
I thought `assert_allclose(actual, expected)` does the same thing, so I deleted it.
I experientially found that this function raises an error if two `dims`'s orders are differ.
Wasn't `assert_allclose` designed to do it?

shoyer replied:
Good point, I missed that this was the xarray function and not the numpy one. So it is indeed redundant (but harmless).

shoyer commented on xarray/core/rolling.py:
```diff
@@ -97,16 +89,55 @@ def __repr__(self):
                                           attrs=','.join(attrs))
 
+    def __len__(self):
+        return len(self.obj[self.dim])
```
Nit: prefer `self.obj.sizes[self.dim]` instead -- it avoids creating a DataArray object (which is somewhat more expensive).

fujiisoup replied:
Thanks.
I'll fix it.

shoyer commented:
This is great -- thanks @fujiisoup !
